### ISSUE TITLE
Extract declarations to other files (including import dependencies)

### ISSUE THREAD

**Issue by angerman** [commercialhaskell/intero#179]
@mwotton's tweet[1](https://twitter.com/mwotton/status/755878392541491200) got me thinking if intero couldn't provide
the facility to move a function (or data declaration or ...) under
selection to a different file, and most importantly take the
relevant import information and place them into the new file,
unless they already exist in that file?

**Comment by chrisdone** [comment]
I've been thinking the same thing. Some kind of "extract" which would put
in your clipboard all the imports and decls necessary for a given selection
of code. That'd be awesome! Intero has the information to do this.

Strawman proposal:

> :extract-at X.hs 1 1 5 1
import Data.List
import Data.Ord
x = sortBy  (comparing fst)

There is also the related notion of extracting any expression and making
all of its missing variables lambda arguments. That could possibly be the
same function.

UX design aside, I'm confirming this is well within Intero's capability to
handle.

If you'd like pointers on tackling it yourself I can write some up.
Basically the ModInfo and it's field modinfoSpans would let you look at all
Ids/Vars in the tree and generate an import list for them.

**Comment by angerman** [comment]
I'd even go a step further (though that might be more part of the actual emacs integration).
Here's the workflow I'd like:

Mark the lines containing the decoration (or even multiple?) and then it e.g.
C-c C-e C-t (extract to), and be prompted for the module / file name.
(would either need to decide, to restrict to only fully contained declarations
or expand the region as necessary to cover all partially covered declarations
at the boundaries.

Next, it would just append the declarations onto that module / file, and
adjust the necessary imports.

On a separate note, if you could reorder declarations like org-mode sections, I could
even go into the new module, once I've extracted all functions into that (new) module
and move them up/down, as I see fit? Though this is probably more haskell-mode
related. Yet I wanted to outline the whole concept.

Thoughts?

**Comment by chrisdone** [comment]
I think we're on the same page. Re-stating in my own language:

0. Select a region which touches some declarations. It doesn't need to select the boundaries exactly, Intero can just see which declarations your selection overlaps.
1. Hit `C-c C-e` (or w/e).
2. Prompt for a filename.
3. Open a buffer with that name and write out the imports and declarations to that buffer.

If the buffer is empty, we just dump the contents in. If the buffer is not empty, then we insert the declarations at the bottom of the file, and the imports at the top after the module header, if any.

I thought that copying to the clipboard would be handy for when I just want to do a Gist or whatever, but you can always write to a throw-away buffer and then delete the file. This keeps the feature's UI simple.

**Comment by tonyday567** [comment]
There's a slightly different UI in my head for this type of thing:

intero-copy: copy decl(s) and necessary imports & pragmas to the clipboard
intero-kill: intero-copy + kill selected decl(s) and no longer needed imports & pragmas
intero-yank: paste haskell code from the clipboard, at point for decl(s) and at the right spot for imports and pragmas.

I'd prefer a clipboard and at point approach if it's possible.  There's less cognitive overload on a big refactor, better control over placement, and you get access to kill-ring and undo functionality.

**Comment by acowley** [comment]
Maybe this should be broken into separate issues to prevent scope creep. Just extracting an expression into a lambda (or function definition syntax) would seem like a really solid goal to me.

In parallel, you've got the notion of lifting an expression and its imports into a new buffer.

Then you've got the combination of the two.

Then you've got sanitation on imports and pragmas. The latter -- removing unneeded pragmas -- is an `hlint` feature these days, iinm.

And we've got the UI design for the whole thing. On that note, I like the idea of opening up in a new buffer to keep things simple, but I'm not convinced we should ask for a file name right off the bat. If the user wants to copy/paste things, the file name is a distraction, and, in the case of intending to create a new module, regular emacs (`C-x C-w`) seems up to the task.

Then again, extracting a sub-expression into a top-level form actually seems to me like the more commonly useful variation of this thing. If it were just me, `C-c C-e` would copy the expression as a lambda to the clipboard, and `C-u C-c C-e` would do the query for file name and create a new buffer thing.

**Comment by angerman** [comment]
What @tonyday567 describes would basically be the building blocks for the move declaration to file / module be, I guess (`intero-kill; switch-buffer buf; intero-yank`). Having those building blocks and then building on top of them would be a rather encompassing UI.

The subexpression point @acowley highlighted, I didn't even consider.  My main focus was on top level expressions. If you had sub level expression support though, you'd certainly have a much more powerful tool though.  I do have no idea if subexpressions would entail much more complexity compared to just handling top-level functions?

**Comment by chrisdone** [comment]
Great discussion so far guys. I'll think more about this too when I'm back
at my desk. I think we're converging on common needs.

I believe as far as the backend is concerned an "extract" could handle
expression extraction to the same module and to a new module, or an
existing module with the same feature. Doing a copy to clipboard would
effectively copy to a new nonexistent module.

From an implementation perspective, I don't think that sub-expressions
represent a significant challenge over declarations in terms of tracking
down dependencies.

**Comment by chrisdone** [comment]
Reducing Intero's scope, closing.
